src/H5Iint.c

```diff
@@ -8811,505 +8866,162 @@ H5I__discard_mt_id_info(H5I_mt_id_info_t * id_info_ptr)
+                        atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_frees_skipped_due_to_empty), 1);
+                        done = TRUE;
+                        break;
+                    }
 
-                    new_fl_stail.ptr = id_info_ptr;
+                    /* attempt to set H5I_mt_g.id_info_fl_stail to fl_snext.  It doesn't
+                     * matter whether we succeed or fail, as if we fail, it
+                     * just means that some other thread beat us to it.
+                     *
+                     * that said, it doesn't hurt to collect stats
+                     */
+                    assert(fl_snext.ptr);
+                    new_fl_stail.ptr = fl_snext.ptr;
                     new_fl_stail.sn  = fl_stail.sn + 1;
-                    if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.id_info_fl_stail), 
-                                                          &fl_stail, new_fl_stail) ) {
+                    if ( ! atomic_compare_exchange_strong(&(H5I_mt_g.id_info_fl_stail), &fl_stail, 
+                                                          new_fl_stail) ) {
 
-                        atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_head_update_cols), 1);
+                        atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_tail_update_cols), 1ULL);
                     }
+                } else if ( ( 0 < (serial_num = atomic_load(&(fl_shead.ptr->serial_num))) ) &&
+                            ( serial_num >= atomic_load(&(H5I_mt_g.id_max_realloc_sn)) ) ) {
 
-                    /* if appropriate, attempt to update H5I_mt_g.max_id_info_fl_len.  In the
-                     * event of a collision, just ignore it and go on, as I don't see any
-                     * reasonable way to recover.
+                    /* if serial_num is zero, it should have already been removed from the free
+                     * list by another thread -- if so, the following attempt to remove if from
+                     * the free list will fail.  If this is not the case, we will catch the error
+                     * after we successfully remove the entry.
                      */
-                    if ( (fl_len = atomic_load(&(H5I_mt_g.id_info_fl_len))) >
-                         (max_fl_len = atomic_load(&(H5I_mt_g.max_id_info_fl_len))) ) {
 
-                        atomic_compare_exchange_strong(&(H5I_mt_g.max_id_info_fl_len), &max_fl_len, fl_len);
-                    }
-
-                    done = true;
+                    /* No reallocable entries available -- just update stats and quit */
+                    atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_frees_skipped_due_to_no_reallocable_entries), 1ULL);
+                    done = TRUE; 
 
                 } else {
 
-                    /* append failed -- update stats and try again */
-                    atomic_fetch_add(&(H5I_mt_g.num_id_info_fl_append_cols), 1);
+                    /* set up new_fl_shead */
```

Comment:
This branch doesn't handle the case where `fl_shead.ptr->serial_num == 0`, even though the if/else explicitly checks for it.

Reply:
If I'm reading the code correctly than this case is known to fail, but is considered a graceful failure as opposed to alternatives. It might be a good idea to explicitly fail gracefully here.